src/csr.tex

```diff
@@ -157,19 +157,23 @@ \section{CSR Instructions}
 
 \subsection*{CSR Access Ordering}
 
-On a given hart, explicit and implicit CSR access are performed in program
-order with respect to those instructions whose execution behavior is affected
-by the state of the accessed CSR. In particular, a CSR access is performed
+Each RISC-V hart normally observes its own CSR accesses, including its
+implicit CSR accesses, as performed in program order.
+In particular, unless specified otherwise, a CSR access is performed
 after the execution of any prior instructions in program order whose behavior
 modifies or is modified by the CSR state and before the execution of any
 subsequent instructions in program order whose behavior modifies or is modified
-by the CSR state. Furthermore, a CSR read access instruction returns the
-accessed CSR state before the execution of the instruction, while a CSR write
-access instruction updates the accessed CSR state after the execution of the
+by the CSR state.
+Any side effects from a CSR access are likewise observed to occur
+synchronously in program order, unless documented otherwise.
```

Comment:
What does "synchronously" mean in this context? Does this imply "atomically" as well (since the side-effect statement was removed below)?

For example, assume there is an interrupt pending and that interrupt is disabled. If a CSR write enables that interrupt, does the architecture require the interrupt to be taken on the next instruction? Or does the architecture imply that is the behavior unless the implementer states otherwise?

Reply:
"Synchronously" is intended to mean that, by default, the full consequences of any side effects are expected to be observable by the very next instruction (to the extent they are observable).  I think the word *atomic* has implications with respect to other, independent observers that I don't intend.

> For example, assume there is an interrupt pending and that interrupt is disabled. If a CSR write enables that interrupt, does the architecture require the interrupt to be taken on the next instruction?

If the interrupt is pending in CSR mip, I've always believed that was an existing requirement.  I would be surprised to learn it's not.

> Or does the architecture imply that is the behavior unless the implementer states otherwise?

That raises a good point.  By "documented" I meant not by an implementor but by the standard(s), assuming we're talking about a standard CSR.

Is "unless specified otherwise" sufficient to plug that hole?  That same crutch already appears in a few places in the documents, so if it's also ambiguous, we have a larger problem.

Reply:
> 
> 
> "Synchronously" is intended to mean that, by default, the full consequences of any side effects are expected to be observable by the very next instruction (to the extent they are observable). I think the word _atomic_ has implications with respect to other, independent observers that I don't intend.

OK. Makes sense.

> 
> > For example, assume there is an interrupt pending and that interrupt is disabled. If a CSR write enables that interrupt, does the architecture require the interrupt to be taken on the next instruction?
> 
> If the interrupt is pending in CSR mip, I've always believed that was an existing requirement. I would be surprised to learn it's not.

Hmmm.... Originally I was thinking that, in general, we didn't want to mix the notions of "synchronous execution" and "side-effects" because doing so would imply that a CSR access may have high synchronization or hazarding costs. So I had been separating the idea of a side-effect, which I considered to be asynchronous, from the simple change of architectural state due to the CSR access that could affect the next instruction in program order. Along those lines, I had been considering that recognizing an interrupt was one such asynchronous side-effect, so that taking a trap wasn't required immediately after enabling a pending interrupt.

Regardless of what further arguments I could make, I understand the new text implies that all side-effects of a CSR access are synchronous unless otherwise allowed by the specification. That seems like a tighter description and clearly leads to your conclusion regarding interrupts in all cases.

> 
> > Or does the architecture imply that is the behavior unless the implementer states otherwise?
> 
> That raises a good point. By "documented" I meant not by an implementor but by the standard(s), assuming we're talking about a standard CSR.
> 
> Is "unless specified otherwise" sufficient to plug that hole? That same crutch already appears in a few places in the documents, so if it's also ambiguous, we have a larger problem.

To me, at least, "specified" concerns itself with the architectural specifications, while "documented" seems broader (but that's probably because I've spent a bit to much time pouring over "product documentation"). My vote would be for "specified."